R/epidatacall.R

```diff
@@ -68,187 +58,213 @@ create_epidata_call <- function(endpoint, params, meta = NULL,
+  response <- request_impl(epidata_call, "classic", fields)
+  response_content <- httr::content(response, as = "text", encoding = "UTF-8")
+  response_content <- jsonlite::fromJSON(response_content, simplifyDataFrame = !disable_data_frame_parsing)
+
+  # success is 1, no results is -2, truncated is 2, -1 is generic error
+  if (response_content$result != 1) {
+    rlang::abort(paste0("epidata error: ", response_content$message), "epidata_error")
+  }
+  if (response_content$message != "success") {
+    rlang::warn(paste0("epidata warning: ", response_content$message), "epidata_warning")
   }
-  m
+  return(response_content$epidata)
 }
 
-#' Fetches the data and returns the CSV text
-#' @rdname epidata_call
+#' Fetches the data and returns a tibble or an `epidata_csv`
 #'
 #' @param epidata_call an instance of `epidata_call`
 #' @param fields filter fields
-#' @importFrom httr stop_for_status content
-#' @importFrom rlang abort
+#' @param disable_date_parsing Boolean. Optionally, `TRUE` to disable parsing of
+#'   columns we expect to be dates, keeping them as character columns instead.
+#'   `FALSE` (the default) to parse these columns into `Date` vectors.
+#' @param disable_tibble_output Boolean. Optionally, `TRUE` to output a
+#'   character vector with the `epidata_csv` class (which provides a custom
+#'   `print` method). `FALSE` (the default) to output a tibble.
 #' @return
-#' - For `fetch_csv`: a string containing CSV text, with the `"epidata_csv"` class added
+#' - For `fetch_csv`: a tibble, or `epidata_csv` if requested with
+#'   `disable_tibble_output = TRUE`
 #'
-#' @export
-fetch_csv <- function(epidata_call, fields = NULL) {
-  stopifnot(inherits(epidata_call, "epidata_call"))
-  stopifnot(is.null(fields) || is.character(fields))
-  if (epidata_call$only_supports_classic) {
-    rlang::abort("the endpoint only supports the classic message format, due to a non-standard behavior",
-      epidata_call = epidata_call,
-      class = "only_supports_classic_format"
-    )
-  }
-  res <- request_impl(epidata_call, "csv", fields)
-  httr::stop_for_status(res)
-  data <- httr::content(res, "text", encoding = "UTF-8")
-  class(data) <- c("epidata_csv", class(data))
-  data
-}
-
-#' Fetches the data in JSON format, then converts into an R object
-#' @rdname epidata_call
-#'
-#' @param epidata_call an instance of `epidata_call`
-#' @param fields filter fields
-#' @param disable_date_parsing disable automatic date parsing
 #' @importFrom httr stop_for_status content
-#' @importFrom jsonlite fromJSON
 #' @importFrom rlang abort
-#' @return
-#' - For `fetch_json`: a `data.frame` form of the object from the parsed JSON
-#'   response
-#'
-#' @export
-fetch_json <- function(epidata_call, fields = NULL, disable_date_parsing = FALSE) {
+fetch_csv <- function(epidata_call, fields = NULL, disable_date_parsing = FALSE, disable_tibble_output = FALSE) {
```

Comment:
question: can we eliminate `fetch_csv` altogether?  Or is it used internally somewhere / otherwise desirable?

Reply:
I think we can. I kept it around just so we had a test reference for the new `fetch_tbl`.